packages/require/lib/typescript.ts

```diff
@@ -0,0 +1,171 @@
+import path from "path";
+
+import type { CreateOptions, Service } from "ts-node";
+import TruffleConfig from "@truffle/config";
+import { readFileSync } from "fs";
+
+import originalRequire from "original-require";
+
+import { Script } from "vm";
+import { TsNodeDependencyError } from "./types";
+
+const _tsExtensions = [".ts", ".tsx", ".cts"];
+
+/**
+ * @param conf the instance of `@truffle/config` for the user's project
+ *
+ * @param sourceFilePath the path to the source module to be compiled
+ *
+ * @param context the global context in which the compiled script will run
+ *
+ * @returns the TypeScript code stored in the file at location `sourceFilePath`
+ *   transpiled to JavaScript when `scriptPath` has a TypeScript file extension,
+ *   otherwise it returns the contents of the `scriptPath` file.
+ */
+export function compile(
+  conf: TruffleConfig,
+  sourceFilePath: string,
+  context: Object
+): string {
+  const source = readFileSync(sourceFilePath, { encoding: "utf-8" });
+
+  // we only compile TS files, so just return the source unless we are dealing
+  // with one of those
+  if (!_tsExtensions.includes(path.extname(sourceFilePath))) {
```

Comment:
Good idea re: regex. I'll change that.

Re: `mts` files, I don't believe that they are a concern, as I don't think that support for EJS targets is considered to be in scope for this change. Please do let me know if you think it should be, though!

Reply:
I just did a bit more research regarding `.mts` files, and I think this should also be regarded as a separate task.

Reviewing the node docs on [differences between ES modules and CommonJS](https://nodejs.org/api/esm.html#differences-between-es-modules-and-commonjs), it would seem that in order to properly support ES modules we'd need to detect whether we're loading such, and in that case we'd need to provide a slightly different global context (e.g. `import` should be present instead of `require`, `__dirname` and `__filename` should not be present, `import.meta.url` should be populated, etc).

Reply:
Ok. Can you make an issue for handling ejs/mts and then resolve this comment thread?

Reply:
Will do.

Reply:
#4328